### ISSUE TITLE
Guidance on using `VoxelGeneratorScript` to generate smooth SDF voxels?

### ISSUE THREAD

**Issue by SummitCollie** [Zylann/godot_voxel#579]
Thanks so much for your making your work available, this module is really cool!

I'm having trouble getting my `VoxelLodTerrain` to generate any visible output. My scene looks like this: ![Scene Tree](https://github.com/Zylann/godot_voxel/assets/41765990/5b4e80cf-c7c8-4a61-9da4-0ebfd335890b)

I'm using a recent Windows build of the editor ([1a51cc2](https://github.com/Zylann/godot_voxel/commit/1a51cc225836d673747b74b77cebbd89ef2e8df8))

My VoxelLodTerrain is set up with the default ShaderMaterial, just added a fragment shader which sets the albedo and left the vertex shader alone for now. This is the main script on the scene root:

```gdscript
extends Node3D

@onready var terrain = $VoxelLodTerrain
const MyGenerator = preload("res://voxel_generator_test/my_generator.gd")

func _ready():
    terrain.generator = MyGenerator.new()
    # My level is bounded so I was hoping to contain the voxel rendering like this,
    # this bounding box looks correct in the editor
    terrain.set_voxel_bounds(AABB(Vector3.ZERO, Vector3.ONE * 512))
```

And `my_generator.gd` looks like:

```gdscript
extends VoxelGeneratorScript

const channel : int = VoxelBuffer.CHANNEL_SDF

func _get_used_channels_mask() -> int:
    return 1 << channel

func _generate_block(buffer : VoxelBuffer, origin : Vector3i, lod : int) -> void:
    if lod != 0: return

    var distance_to_zero = Vector3(origin).distance_to(Vector3.ZERO)
    if distance_to_zero > 100 && distance_to_zero < 150:
	  buffer.fill_f(1, channel)
    else:
	  buffer.fill_f(0, channel)
```

And... nothing renders. I'm pretty new to Godot and this module so it's quite possible I'm missing an important step or doing something dumb, but the docs and sample projects seem to be lacking examples of how to make a generator for SDF `VoxelLodTerrain` so I'm stuck.

The script above is my attempt to just produce any visible result, but my ultimate goal is to generate a solid 3D block of voxels and carve tunnels out of it based on other logic in my game code, like distance from a Curve3D (inside the radius of the tunnel = no voxels). IDK whether this is even a sane approach but it seems like a better option than using VoxelTool to carve tunnels at runtime.

Any advice is greatly appreciated!

**Comment by Zylann** [comment]
> And... nothing renders.

In the editor? Or did you launch the game?
Don't forget to have a `VoxelViewer` in your scene, as this is necessary to tell the terrain where to stream chunks.

**Comment by SummitCollie** [comment]
> In the editor? Or did you launch the game?

Ah sorry I should've clarified since I know people frequently neglect to do that stuff: I initially used the flat generator as a sanity check and I *did* see its output using the same setup (both in editor & in game). The camera (with VoxelViewer child) is moveable with WASD/mouse via a script, and I always check by zooming and rotating to different angles.

I'm also aware that my generator script's output won't be visible in the editor unless I enable the "Run Stream In Editor" option on VoxelLodTerrain, which I shouldn't do unless I'm planning not to change any script files (because the editor accessing them would cause problems).

> Signed distances need to cross 0 in order to produce a surface. All your voxels here will either have distance 0 or 1, so never crossing zero. So no surface will appear. At least maybe you need to set -1 instead of 0.

Ah yeah thanks, I think this is the concept I wasn't understanding, even having read over your SDF doc (perhaps not thoroughly enough).

> > if lod != 0: return
> And this pretty much prevents you to see any of it as well.

Okay, that makes sense. I copied my script from the example I found in [the scripting docs](https://voxel-tools.readthedocs.io/en/latest/scripting/#example), I guess it works in that scenario because blocky voxels don't support LOD therefore always have `lod == 0`? But I was curious about how anything outside the first LOD zone would get rendered with that line there... figured I should at least see some stuff near the camera though.

Thanks for all your help, and appreciate the example, I'll have to play with that after work!

If you feel like indulging me further, the main thing I'm still wondering how I'll accomplish is applying noise to the interior walls of the tunnels I carve (to make them look like caves, maybe add stuff stalactites etc).

I need a guaranteed minimum clearance between cave walls (as well as known start/end points) which is why I'm not just using a noise algo for the whole thing. Your [procedural generation docs on determinism](https://voxel-tools.readthedocs.io/en/latest/procedural_generation/#deterministic-approach) seem like a good starting point but if you've got any other thoughts/resources on that topic I'd be happy to hear/learn of them!

Closing since you've given me plenty to work with and it's my fault if I can't get it to render at this point lol.
